test/facility/put-facility-by-facility-identifier.api-test.ts

```diff
@@ -0,0 +1,197 @@
+import { UpdateFacilityGenerator } from '@ukef-test/support/generator/update-facility-generator';
+import nock from 'nock';
+import supertest from 'supertest';
+
+import { withAcbsGetFacilityServiceCommonTests } from '../common-tests/acbs-get-facility-by-identifier-api-tests';
+import { withAcbsUpdateFacilityByIdentifierServiceTests } from '../common-tests/acbs-update-facility-by-identifier-api-tests';
+import { PutFacilityAcbsEndpoints } from '../support/interfaces/put-facility-by-facility-identifier-acbs-endpoints.interface.';
+
+const testCases = [ENUMS.FACILITY_UPDATE_OPERATIONS.ISSUE, ENUMS.FACILITY_UPDATE_OPERATIONS.AMEND_EXPIRY_DATE];
+
+testCases.forEach((queryToTest) => {
+  describe(`PUT /facilities?op=${queryToTest}`, () => {
+    let api: Api;
+
+    beforeAll(async () => {
+      api = await Api.create();
+    });
+
+    afterAll(async () => {
+      await api.destroy();
+    });
+
+    afterEach(() => {
+      nock.abortPendingRequests();
+      nock.cleanAll();
+    });
+
+    const valueGenerator = new RandomValueGenerator();
+    const dateStringTransformations = new DateStringTransformations();
+    const portfolioIdentifier = PROPERTIES.GLOBAL.portfolioIdentifier;
+    const facilityIdentifier = valueGenerator.facilityId();
+
+    const updateFacilityBaseUrl = `/api/v1/facilities/${facilityIdentifier}`;
+    const updateFacilityUrl = updateFacilityBaseUrl + `?op=${queryToTest}`;
+
+    const {
+      updateFacilityRequest,
+      acbsGetExistingFacilityResponse,
+      acbsUpdateFacilityRequest: expectedAcbsUpdateFacilityRequest,
+    } = new UpdateFacilityGenerator(valueGenerator, dateStringTransformations).generate({ numberToGenerate: 1, facilityIdentifier });
+
+    const expectedPutResponse = { facilityIdentifier: facilityIdentifier };
+
+    withClientAuthenticationTests({
+      givenTheRequestWouldOtherwiseSucceed: () => requestToUpdateFacility(),
+      makeRequestWithoutAuth: () => makeRequestWithoutAuth(),
+    });
+
+    if (queryToTest === ENUMS.FACILITY_UPDATE_OPERATIONS.ISSUE) {
+      const unissuedFacilityStageCode = TEST_FACILITY_STAGE_CODE.unissuedFacilityStageCode;
+      it('returns a 400 response if request has an unissued facility stage code"', async () => {
+        givenTheRequestWouldOtherwiseSucceed();
+
+        const restult = givenRequestToGetFacilityInAcbsSucceeds();
+        restult.isDone();
+        const modifiedUpdateFacilityRequest = { ...updateFacilityRequest, facilityStageCode: unissuedFacilityStageCode };
+
+        const { status, body } = await makeRequestWithBody(modifiedUpdateFacilityRequest);
+
+        expect(status).toBe(400);
+        expect(body).toStrictEqual({ message: 'Bad request', error: 'Facility stage code is not issued', statusCode: 400 });
+      });
+    }
+
+    it('returns a 200 response with the facility identifier if the facility has been successfully updated in ACBS', async () => {
```

Comment:
I think it's nicer in terms of readability if the happy path test is at the top and the error cases follow

Reply:
Good shout